### PR TITLE
DEVREL-977 Handle circular data

### PR DESCRIPTION
**Pull request by pokornyd**
### Motivation

[DEVREl-977](https://kontent-ai.atlassian.net/jira/software/projects/DEVREL/boards/421?selectedIssue=DEVREL-977)

Adds handler for sanitizing circular references.

### Checklist

- [ ] Code follows coding conventions held in this repo
- [ ] Automated tests have been added
- [ ] Tests are passing
- [ ] Docs have been updated (if applicable)
- [ ] Temporary settings (e.g. variables used during development and testing) have been reverted to defaults

### How to test

Introduce a circular reference in Kontent.ai project (e.g. via content item link in any given Fact item), run the app and visit any page that requests said item.

### CODE REVIEW

**Comment by IvanKiral** [comment]
build is falling, please fix it first :)

File: lib/utils/circularityUtils.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+  circularReferences: CircularReferenceInfo[] = []
+): [TItem, CircularReferenceInfo[]] => {
+  if (seenCodenames.has(data.system.codename)) {
+    const seenCodenamesArray = Array.from(seenCodenames);
+    const itemCodename = seenCodenamesArray[seenCodenamesArray.length - 1]!;
+    return [
+      // Replaces circular reference with null. Cast necessary to adhere to type signature.
+      null as unknown as TItem,
+      [
+        ...circularReferences,
```
Comment (JiriLojda) [inline lib/utils/circularityUtils.ts:57]: If you'd return just the newfound reference (not all the references that the caller knows already anyway), you wouldn't need to filter all the old references out on line `88`.  Btw. the filtering itself is of quadratic complexity run for each linked item element of all items.

File: lib/utils/circularityUtils.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  seenCodenames.add(data.system.codename);
+
+  const sanitizedElements: Record<
+    string,
+    ContentItemElementsIndexer | LinkableElement
+  > = {};
+
+  // Recursively run the method for rich text and linked item elements, leave other elements unchanged.
+  for (const [elementCodename, element] of Object.entries(data.elements)) {
+    sanitizedElements[elementCodename] = isLinkableElement(element)
```
Comment (JiriLojda) [inline lib/utils/circularityUtils.ts:76]: This mutation wouldn't be necessary if you'd used the `.map` function instead of the `foreach` cycle leading to imho more comprehensible code.

File: pages/[envId]/products/index.tsx [hunk 3]
```diff
@@ -214,12 +217,31 @@ export const getStaticProps: GetStaticProps<Props, { envId: string }> = async co
...
+  }
+
+  let productsCircularReferences: ItemCircularReferenceMap = {};
+
+  const products = {
+    ...productsData,
+    items: productsData.items.map(product => {
+      const [sanitizedProduct, foundCycles] = sanitizeCircularData(product);
+      productsCircularReferences = {...productsCircularReferences, ...foundCycles};
+      return sanitizedProduct;
```
Comment (JiriLojda) [inline pages/[envId]/products/index.tsx:235]: Handling an array of products seems to be common enough to justify a utility function, it is quite a chunk of code.

Comment (JiriLojda) [reply pages/[envId]/products/index.tsx:235]: btw. it can be done functionally quite eloquently without a mutating object. 😝

File: pages/callback.tsx [hunk 4]
```diff
@@ -107,7 +107,7 @@ const CallbackPage: React.FC = () => {
         return setError(api_key.error);
       }
 
-      window.location.replace(authResult.appState ?? '/'); // router.replace changes the "slug" query parameter so we can't use it here, because this parameter is used when calling the /api/preview endpoint
+      window.location.replace(getCookie(urlAfterAuthCookieName) ?? '/'); // router.replace changes the "slug" query parameter so we can't use it here, because this parameter is used when calling the /api/preview endpoint
```
Comment (JiriLojda) [inline pages/callback.tsx:110]: This should already be in the `main` branch. Just rebase on it.

**Review by JiriLojda — CHANGES_REQUESTED** [review]
A few suggestions from a quick look.

File: lib/utils/circularityUtils.ts (hunk 1 above)
Comment (pokornyd) [reply lib/utils/circularityUtils.ts:71]: good point, I somehow managed to make this part more complicated while trying to simplify the rest ... refactored accordingly

File: pages/[envId]/products/index.tsx (hunk 3 above)
Comment (pokornyd) [reply pages/[envId]/products/index.tsx]: products don't actually have any linked item elements or rich text, so I removed the sanitation from the page. there are however potential linked items in article listing, so I added an extra method. I decided not to reuse the existing one because I felt it would add time complexity in its current state. some further refactor may be in order...

File: lib/utils/circularityUtils.ts (hunk 2 above)
Comment (pokornyd) [reply lib/utils/circularityUtils.ts]: refactored

File: pages/callback.tsx (hunk 4 above)
Comment (pokornyd) [reply pages/callback.tsx]: rebased

File: lib/utils/circularityUtils.ts [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+ * Performs a depth-first traversal to sanitize any content item references that would create a cycle.
+ * Collects information on identified circular references.
+ * @param data - The content item to sanitize.
+ * @param elementName - The name of the current element being processed.
+ * @param seenCodenames - Set of content item codenames visited during the recursion.
+ * @param circularReferences - Accumulated list of detected circular references.
+ * @returns A tuple containing the sanitized content item and the accumulated circular references.
+ */
+const sanitizeCircularDataInternal = <TItem extends IContentItem>(
+  data: TItem,
```
Comment (IvanKiral) [inline lib/utils/circularityUtils.ts:57]: why is this called `data` and not `item`?

File: lib/utils/circularityUtils.ts [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+  elementName = "",
+  seenCodenames = new Set<string>()
+): [TItem, CircularReferenceInfo[]] => {
+  const circularReferences: CircularReferenceInfo[] = [];
+
+  if (seenCodenames.has(data.system.codename)) {
+    const seenCodenamesArray = Array.from(seenCodenames);
+    const itemCodename = seenCodenamesArray[seenCodenamesArray.length - 1]!;
```
Comment (IvanKiral) [inline lib/utils/circularityUtils.ts:65]: why not parentItemCodename?

File: lib/utils/circularityUtils.ts [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+/**
+ * Sanitizes a content item by removing references that would otherwise create a cycle, potentially crashing the app.
+ * Collects information on sanitized references (grouped by item codename) for use by warning messages.
+ * @param item - The content item to sanitize.
+ * @returns A tuple containing the sanitized content item and a structured record of detected circular references.
+ */
+export const sanitizeItem = <TItem extends IContentItem>(
+  item: TItem
+): [TItem, ItemCircularReferenceMap] => {
+  const [sanitizedData, circularInfoArray] = sanitizeCircularDataInternal(item);
```
Comment (IvanKiral) [inline lib/utils/circularityUtils.ts:119]: sanitiziedItem?

**Review by IvanKiral — CHANGES_REQUESTED** [review]
The code seems to be right, I would make some changes to the naming. Also I would consider writing a few of the tests for those functions :)

File: lib/utils/circularityUtils.ts [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+ */
+export const sanitizeItemListing = <TItem extends IContentItem>(
+  itemListing: IContentItemListingResponse<TItem>
+): [IContentItemListingResponse<TItem>, ItemCircularReferenceMap] => {
+  const listingCircularInfoArray: CircularReferenceInfo[] = [];
+  const sanitizedItems: TItem[] = [];
+
+  itemListing.items.forEach((item) => {
+    const [sanitizedItem, circularInfoArray] =
+      sanitizeCircularDataInternal(item);
```
[the remark this answers is not in the document]
Comment (pokornyd) [reply lib/utils/circularityUtils.ts:151]: autoformatting in VScode via prettier or dprint, not sure which one I use for TS :/

File: lib/utils/circularityUtils.ts (hunk 5 above)
Comment (pokornyd) [reply lib/utils/circularityUtils.ts:57]: oversight when renaming :) fixed

File: lib/utils/circularityUtils.ts [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+
+  if (seenCodenames.has(data.system.codename)) {
+    const seenCodenamesArray = Array.from(seenCodenames);
+    const itemCodename = seenCodenamesArray[seenCodenamesArray.length - 1]!;
+
+    return [
+      // Replaces circular reference with null. Cast necessary to adhere to type signature.
+      null as unknown as TItem,
+      [
+        ...circularReferences,
```
[the remark this answers is not in the document]
Comment (pokornyd) [reply lib/utils/circularityUtils.ts:71]: also leftover from previous implementation, good catch, removed

File: lib/utils/circularityUtils.ts (hunk 7 above)
Comment (pokornyd) [reply lib/utils/circularityUtils.ts:119]: renamed both the method and return value

File: lib/utils/circularityUtils.ts (hunk 6 above)
Comment (pokornyd) [reply lib/utils/circularityUtils.ts:65]: fixed

**Review by IvanKiral — APPROVED** [review]

**Comment by Enngage** [comment]
Hey guys, just noticed this PR. Why do you have a need to handle circular references? From my experience this can happen if you pass data from server components to client components (next.js app router), but this would be an anti-pattern because you should only pass the data that is needed by client component (as opposed to a full response) to reduce the transmitted size & improve performance. 

If you need to serialize objects with circular references in backend, you could use something like https://www.npmjs.com/package/flatted

And it is also possible to disable linked items mapping to prevent circular references from ever occuring (https://github.com/kontent-ai/delivery-sdk-js#handling-circular-references) - this however means you need to get the items by codename manually.

**Comment by pokornyd** [comment]
> Hey guys, just noticed this PR. Why do you have a need to handle circular references? From my experience, this can happen if you pass data from server components to client components (next.js app router), but this would be an anti-pattern because you should only pass the data that is needed by client component (as opposed to a full response) to reduce the transmitted size & improve performance.
> 
> If you need to serialize objects with circular references in the backend, you could use something like https://www.npmjs.com/package/flatted
> 
> It is also possible to disable linked items mapping to prevent circular references from ever occurring (https://github.com/kontent-ai/delivery-sdk-js#handling-circular-references) - this however means you need to get the items by codename manually.
> 
> Sanitizing circular objects like in this PR kinds of defeats the purpose of having linked items because you "never really know" if you aren't going to need to use/reference the linked items and your call might fail because the linked item has been sanitized, but does actually exist. You would have to check whether the item has been sanitized to ensure you retrieve the correct data.

as you've mentioned, the data fetching with Next.js revolves around JSON, which is not friends with circular data. While building the individual projects, we tried to reuse some of the existing components, to realize it can close a cycle somewhere else, effectively crashing the app. 

I agree the the app should pass only the necessary data, this is why the methods also collect information on the items that were sanitized, subsequently passing them to the root component, which displays warning for each sanitized item, including all identified cycles and elements they were found in. The warning also suggest revisiting the data structure or modifying the depth parameter, thus pushing you towards only passing the actually required data.

In simpler terms, we basically caught the generic Next error message and made it more verbose to make it easier to act upon, as some users may introduce circular references when modifying the sample project.

Flatted would certainly be applicable but we'd either need to deserialize the response back to the strongly typed model on client side or adjust the props and logic on all affected components, unless I'm mistaken.

I can modify the warning message to also mention the possibility of disabling the mapping in the SDK, or add any further "best practice" mentions.

**Comment by pokornyd** [comment]
following internal discussion, this approach was deemed unnecessary, abandoned in favor of flattening: https://github.com/kontent-ai/sample-app-next-js/pull/57
